### ISSUE TITLE
visit_bytes / visit_byte_buf

### ISSUE THREAD

**Issue by arcnmx** [serde-rs/serde#398]
`de::Visitor::visit_bytes` isn't implemented for `Vec<u8>` or `Box<[u8]>`, which makes this method a bit of a hidden mine... If a (de)serialization implementation tries supporting/calling it, there will be no way to actually interact with the values (unless a custom newtype is created and used) and you'll just get this error:

    Invalid type. Expected `Bytes`

This is probably an issue blocked on specialization (because of conflicts with the blanket impls on `Vec<T>`) but perhaps some sort of mitigation in the meantime would be helpful...
- Deprecate / remove / note in the documentation that these methods shouldn't be used yet. This means that crates will need to be manually updated later once the functionality actually becomes available.
- Provide an official newtype wrapper? Maybe used in combination with the new serializer attributes.
- Change the default impl to fallback to a slower seq implementation rather than error outright. This seems like the best migration path, but may cause weirdness like allowing byte buffers to deserialize into any numeric array type. That behaviour probably isn't any worse than what we're stuck with doing now, though.

Also slightly unrelated, but `String` only impls `visit_bytes` and not `visit_byte_buf`, which should be implemented via `String::from_utf8`.

**Comment by dtolnay** [comment]
Here is `visit_byte_buf` for String: [`StringVisitor::visit_byte_buf`](https://github.com/serde-rs/serde/blob/5fb73073bd14669ef3a5fd7eeec69ed5fe12d1ea/serde/src/de/impls.rs#L295-L302).

Here is the official wrapper corresponding to `Vec<u8>`: [`serde::bytes::ByteBuf`](https://docs.serde.rs/serde/bytes/struct.ByteBuf.html). There is also [`serde::bytes::Bytes`](https://docs.serde.rs/serde/bytes/struct.Bytes.html) corresponding to `&[u8]` which supports serialization only.

**Comment by dtolnay** [comment]
Closing as a duplicate of #309.
